# Data Intelligence Exam

Alexander lolk | lolk@itu.dk | Date: 26-05-2026  

## Introduction

Here the student gives an introduction to the dataset and some social and cultural context.

They give an overview of the report the have written and the different steps they will go through.

## Loading and overview

Here the student starts explaining what the different steps do and why they use the individual blocks of code.

In [2]:
%pip install -U pandas 
%pip install -U matplotlib 
%pip install -U seaborn 
%pip install -U scikit-learn

     ---------------------------------------- 0.0/60.8 kB ? eta -:--:--
     ------------------- ------------------ 30.7/60.8 kB 660.6 kB/s eta 0:00:01
     ---------------------------------------- 60.8/60.8 kB 1.1 MB/s eta 0:00:00
   ---------------------------------------- 0.0/11.6 MB ? eta -:--:--
   - -------------------------------------- 0.5/11.6 MB 14.9 MB/s eta 0:00:01
   ----- ---------------------------------- 1.6/11.6 MB 20.3 MB/s eta 0:00:01
   ---------- ----------------------------- 3.2/11.6 MB 25.4 MB/s eta 0:00:01
   ----------------- ---------------------- 5.2/11.6 MB 30.4 MB/s eta 0:00:01
   ----------------------- ---------------- 6.9/11.6 MB 31.4 MB/s eta 0:00:01
   ------------------------------ --------- 8.8/11.6 MB 33.2 MB/s eta 0:00:01
   ------------------------------------ --- 10.7/11.6 MB 36.4 MB/s eta 0:00:01
   ---------------------------------------  11.6/11.6 MB 38.6 MB/s eta 0:00:01
   ---------------------------------------- 11.6/11.6 MB 34.4 MB/s eta 0


[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\Alexander\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


     ---------------------------------------- 0.0/106.6 kB ? eta -:--:--
     ----------- --------------------------- 30.7/106.6 kB 1.4 MB/s eta 0:00:01
     -------------------------------------- 106.6/106.6 kB 2.0 MB/s eta 0:00:00
   ---------------------------------------- 0.0/8.1 MB ? eta -:--:--
   -- ------------------------------------- 0.6/8.1 MB 19.2 MB/s eta 0:00:01
   -------- ------------------------------- 1.6/8.1 MB 20.7 MB/s eta 0:00:01
   ---------------- ----------------------- 3.3/8.1 MB 26.0 MB/s eta 0:00:01
   ------------------------- -------------- 5.1/8.1 MB 29.8 MB/s eta 0:00:01
   ---------------------------------- ----- 7.0/8.1 MB 32.1 MB/s eta 0:00:01
   ---------------------------------------  8.1/8.1 MB 32.2 MB/s eta 0:00:01
   ---------------------------------------  8.1/8.1 MB 32.2 MB/s eta 0:00:01
   ---------------------------------------- 8.1/8.1 MB 24.6 MB/s eta 0:00:00
   ---------------------------------------- 0.0/222.0 kB ? eta -:--:--
   --------


[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\Alexander\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


   ---------------------------------------- 0.0/294.9 kB ? eta -:--:--
   ---- ----------------------------------- 30.7/294.9 kB 1.4 MB/s eta 0:00:01
   ------------------ --------------------- 133.1/294.9 kB 2.0 MB/s eta 0:00:01
   ---------------------------------------- 294.9/294.9 kB 3.6 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\Alexander\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


     ---------------------------------------- 0.0/60.8 kB ? eta -:--:--
     ------------------------- ------------ 41.0/60.8 kB 991.0 kB/s eta 0:00:01
     ---------------------------------------- 60.8/60.8 kB 1.6 MB/s eta 0:00:00
   ---------------------------------------- 0.0/11.1 MB ? eta -:--:--
    --------------------------------------- 0.2/11.1 MB 5.0 MB/s eta 0:00:03
   ---- ----------------------------------- 1.2/11.1 MB 15.3 MB/s eta 0:00:01
   ---------- ----------------------------- 2.9/11.1 MB 22.9 MB/s eta 0:00:01
   ----------------- ---------------------- 4.9/11.1 MB 28.3 MB/s eta 0:00:01
   ------------------------ --------------- 6.9/11.1 MB 31.2 MB/s eta 0:00:01
   ------------------------------- -------- 8.7/11.1 MB 32.6 MB/s eta 0:00:01
   ------------------------------------- -- 10.4/11.1 MB 36.4 MB/s eta 0:00:01
   ---------------------------------------  11.1/11.1 MB 36.4 MB/s eta 0:00:01
   ---------------------------------------- 11.1/11.1 MB 34.6 MB/s eta 0:


[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\Alexander\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


So let's have a look at this dataset.
We start with reading the census csv file and then print it out to look at.

In [101]:
# import pandas libary for data handling
import pandas as pd

# read the csv and puts in a pandas dataframe
df = pd.read_csv("census-1994.csv")

df

age          workclass    education       marital_status  \
0       39          State-gov    Bachelors        Never-married   
1       50   Self-emp-not-inc    Bachelors   Married-civ-spouse   
2       38            Private      HS-grad             Divorced   
3       53            Private         11th   Married-civ-spouse   
4       28            Private    Bachelors   Married-civ-spouse   
...    ...                ...          ...                  ...   
32556   27            Private   Assoc-acdm   Married-civ-spouse   
32557   40            Private      HS-grad   Married-civ-spouse   
32558   58            Private      HS-grad              Widowed   
32559   22            Private      HS-grad        Never-married   
32560   52       Self-emp-inc      HS-grad   Married-civ-spouse   

               occupation    relationship    race   gender  capital_gain  \
0            Adm-clerical   Not-in-family   White     Male          2174   
1         Exec-managerial         Husband   White     Male             0   
2       Handlers-cleaners   Not-in-family   White     Male             0   
3       Handlers-cleaners         Husband   Black     Male             0   
4          Prof-specialty            Wife   Black   Female             0   
...                   ...             ...     ...      ...           ...   
32556        Tech-support            Wife   White   Female             0   
32557   Machine-op-inspct         Husband   White     Male             0   
32558        Adm-clerical       Unmarried   White   Female             0   
32559        Adm-clerical       Own-child   White     Male             0   
32560     Exec-managerial            Wife   White   Female         15024   

       capital_loss  hours_per_week  native_country income_bracket  
0                 0              40   United-States          <=50K  
1                 0              13   United-States          <=50K  
2                 0              40   United-States          <=50K  
3                 0              40   United-States          <=50K  
4                 0              40            Cuba          <=50K  
...             ...             ...             ...            ...  
32556             0              38   United-States          <=50K  
32557             0              40   United-States           >50K  
32558             0              40   United-States          <=50K  
32559             0              20   United-States          <=50K  
32560             0              40   United-States           >50K  

[32561 rows x 13 columns]

In this dataset, we see that it contains some general information about people and their living conditions. \
For now, there does not seem to be any sensitive data to keep an eye for. 

Let see how pandas handles this dataset. 

In [ ]:
# show the column in the dataframe as well as their type.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   education       32561 non-null  object
 3   marital_status  32561 non-null  object
 4   occupation      32561 non-null  object
 5   relationship    32561 non-null  object
 6   race            32561 non-null  object
 7   gender          32561 non-null  object
 8   capital_gain    32561 non-null  int64 
 9   capital_loss    32561 non-null  int64 
 10  hours_per_week  32561 non-null  int64 
 11  native_country  32561 non-null  object
 12  income_bracket  32561 non-null  object
dtypes: int64(4), object(9)
memory usage: 3.2+ MB


Nothing too out of the orinary, what could be number seem to be. \
For now the data looks to be in a fine state.

# Exploration
Now lets explore what is contain within these columns and rows. 

First order of business is to look that the column which was of the int64.
For this we use summary statistics, so as to get overview of the columns. 

In [ ]:
# gives descriptive statistics over all columns that is a type of number.
df.describe()

age  capital_gain  capital_loss  hours_per_week
count  32561.000000  32561.000000  32561.000000    32561.000000
mean      38.581647   1077.648844     87.303830       40.437456
std       13.640433   7385.292085    402.960219       12.347429
min       17.000000      0.000000      0.000000        1.000000
25%       28.000000      0.000000      0.000000       40.000000
50%       37.000000      0.000000      0.000000       40.000000
75%       48.000000      0.000000      0.000000       45.000000
max       90.000000  99999.000000   4356.000000       99.000000

So, when looking a that the "count" part, it is worth noting that it counts all the non-na there exist in the column. \
Given that the count is for the size that the dataset is of, we know that there isn't any worries with non-na in these columns.

As for mean and std, For the columns capital_grain and loss, their standard deviation has a big spread compared to their mean. \
So it would seem with that, that there exist some big outliers.

While for the min, 25%, 50%, 75% and max, here there appear to be also something wrong. 
We can see here that for two columns, capital_gain and loss, there isn't any gain or loss before the last 25% of the data.
This require further investigation.

In [ ]:
# Checks if a row within capital_gain is greater than 0, if so show it.
df[df["capital_gain"] > 0]

age          workclass     education       marital_status  \
0       39          State-gov     Bachelors        Never-married   
8       31            Private       Masters        Never-married   
9       42            Private     Bachelors   Married-civ-spouse   
59      30            Private       HS-grad   Married-civ-spouse   
60      30            Private     Bachelors   Married-civ-spouse   
...    ...                ...           ...                  ...   
32515   66        Federal-gov          10th   Married-civ-spouse   
32518   57          Local-gov       HS-grad   Married-civ-spouse   
32538   38            Private     Bachelors             Divorced   
32548   65   Self-emp-not-inc   Prof-school        Never-married   
32560   52       Self-emp-inc       HS-grad   Married-civ-spouse   

               occupation    relationship    race   gender  capital_gain  \
0            Adm-clerical   Not-in-family   White     Male          2174   
8          Prof-specialty   Not-in-family   White   Female         14084   
9         Exec-managerial         Husband   White     Male          5178   
59      Machine-op-inspct         Husband   White     Male          5013   
60                  Sales         Husband   White     Male          2407   
...                   ...             ...     ...      ...           ...   
32515        Craft-repair         Husband   White     Male          3471   
32518        Craft-repair         Husband   White     Male         99999   
32538      Prof-specialty       Unmarried   Black   Female         15020   
32548      Prof-specialty   Not-in-family   White     Male          1086   
32560     Exec-managerial            Wife   White   Female         15024   

       capital_loss  hours_per_week  native_country income_bracket  
0                 0              40   United-States          <=50K  
8                 0              50   United-States           >50K  
9                 0              40   United-States           >50K  
59                0              40   United-States          <=50K  
60                0              40   United-States          <=50K  
...             ...             ...             ...            ...  
32515             0              40   United-States          <=50K  
32518             0              40   United-States           >50K  
32538             0              45   United-States           >50K  
32548             0              60   United-States          <=50K  
32560             0              40   United-States           >50K  

[2712 rows x 13 columns]

A fairly small chunk of the data, 2712 rows, has a value a in the capital_gain, meaning alot of the data has 0 or less in capital gain. \
worth keep in mind for latter. 

In [50]:
df[df["capital_loss"] > 0]

age          workclass      education       marital_status  \
23      43            Private           11th   Married-civ-spouse   
32      45            Private      Bachelors             Divorced   
52      47            Private    Prof-school   Married-civ-spouse   
93      30            Private        HS-grad   Married-civ-spouse   
96      48   Self-emp-not-inc      Doctorate   Married-civ-spouse   
...    ...                ...            ...                  ...   
32441   53            Private   Some-college   Married-civ-spouse   
32443   18            Private        HS-grad        Never-married   
32445   61            Private        HS-grad              Widowed   
32458   36            Private      Bachelors   Married-civ-spouse   
32500   34            Private           11th             Divorced   

              occupation    relationship                 race   gender  \
23      Transport-moving         Husband                White     Male   
32       Exec-managerial       Own-child                White     Male   
52        Prof-specialty            Wife                White   Female   
93                 Sales            Wife   Asian-Pac-Islander   Female   
96        Prof-specialty         Husband                White     Male   
...                  ...             ...                  ...      ...   
32441   Transport-moving         Husband                White     Male   
32443              Sales       Own-child                White   Female   
32445       Craft-repair   Not-in-family                Black   Female   
32458     Prof-specialty         Husband                White     Male   
32500   Transport-moving       Unmarried                White     Male   

       capital_gain  capital_loss  hours_per_week  native_country  \
23                0          2042              40   United-States   
32                0          1408              40   United-States   
52                0          1902              60        Honduras   
93                0          1573              35               ?   
96                0          1902              60   United-States   
...             ...           ...             ...             ...   
32441             0          1887              45   United-States   
32443             0          1602              20   United-States   
32445             0          1669              50   United-States   
32458             0          1977              45   United-States   
32500             0           880              60   United-States   

      income_bracket  
23             <=50K  
32             <=50K  
52              >50K  
93             <=50K  
96              >50K  
...              ...  
32441           >50K  
32443          <=50K  
32445          <=50K  
32458           >50K  
32500          <=50K  

[1519 rows x 13 columns]

Even less than with capital_gain, it would seem that these two columns have data. \
Since capital gain and loss refers to buying and selling, it would make sense that people how had anything in these columns, would have some sort of occupation that would relate. \
Lets put that hold for now.

In [105]:
df.groupby("age")["hours_per_week"].mean()

age
17    21.367089
18    25.912727
19    30.678371
20    32.280212
21    34.034722
        ...    
85    29.333333
86    40.000000
87     2.000000
88    40.000000
90    36.813953
Name: hours_per_week, Length: 73, dtype: float64

It would seem that the native_country columns have a or more unanswered value, if you look at the last print out.

In [60]:
# prints out all the unique value that exist within the dataframe
df["native_country"].unique()

array([' United-States', ' Cuba', ' Jamaica', ' India', ' ?', ' Mexico',
       ' South', ' Puerto-Rico', ' Honduras', ' England', ' Canada',
       ' Germany', ' Iran', ' Philippines', ' Italy', ' Poland',
       ' Columbia', ' Cambodia', ' Thailand', ' Ecuador', ' Laos',
       ' Taiwan', ' Haiti', ' Portugal', ' Dominican-Republic',
       ' El-Salvador', ' France', ' Guatemala', ' China', ' Japan',
       ' Yugoslavia', ' Peru', ' Outlying-US(Guam-USVI-etc)', ' Scotland',
       ' Trinadad&Tobago', ' Greece', ' Nicaragua', ' Vietnam', ' Hong',
       ' Ireland', ' Hungary', ' Holand-Netherlands'], dtype=object)

For the most part, the structure seem fine with some outliers. But one big problem is that all value has a whitespace in fount of the name.

After some exploration, it would seem that all the columns that is not of the type int64 has a problem with whitespaces. \
This was found out with the method as shown above.

Lets fix these columns in the preprocessing
- workclass
- native_country (whitespace and ?)
- education (whitespace and ?)
- gender (whitespace and ?)
- income_bracket (whitespace)
- marital_status (whitespace)
- occupation (whitespace and ?)
- race (whitespace and weird values)
- relationship (whitespace)


# Research questions and hypothesis

On the basis of the initial data exploration, the student now formulates some questions to ask and some hypotheses to test. \
Does your social situation change your income?

## Preprocessing

First I need to clean the data to make sure the values are all valid. 

In [95]:
# import regular expressions
import re

# this expression checks if the start of a line has a whitespace
reg = "^\s"

# function for checking whitespaces and removing them
def remove_whitespaces(str):
    # checks the column value with regex, if there is match it split it, meaning " hello world" would become [" ", "hello world"]
    mat = re.split(reg, str)
    
    # if the size of the split result in a list with a size of 2, meaning like above, it will then return the part that doesn't have a whitespace
    # else it return back the value as it was 
    if len(mat) == 2:
        return mat[1]
    else:
        return str

# remove_whitespaces(" hello world")

In [103]:
df["education"] = df["education"].map(remove_whitespaces)
df["gender"] = df["gender"].map(remove_whitespaces)
df["income_bracket"] = df["income_bracket"].map(remove_whitespaces)
df["marital_status"] = df["marital_status"].map(remove_whitespaces)
df["native_country"] = df["native_country"].map(remove_whitespaces)
df["occupation"] = df["occupation"].map(remove_whitespaces)
df["race"] = df["race"].map(remove_whitespaces)
df["relationship"] = df["relationship"].map(remove_whitespaces)
df["relationship"].unique()

array(['Not-in-family', 'Husband', 'Wife', 'Own-child', 'Unmarried',
       'Other-relative'], dtype=object)

In [ ]:
df["workclass"] = df["workclass"].map(remove_whitespaces)
df["workclass"].unique()

In [82]:
df["occupation"].unique()

array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Sales', ' Craft-repair',
       ' Transport-moving', ' Farming-fishing', ' Machine-op-inspct',
       ' Tech-support', ' ?', ' Protective-serv', ' Armed-Forces',
       ' Priv-house-serv'], dtype=object)